site/content/posts/2023-08-08-v0.25.0-impersonation-proxy-with-external-certs.md

```diff
@@ -0,0 +1,318 @@
+
+```shell
+kubectl get secret root-secret \
+    --namespace cert-manager \
+    --output jsonpath="{.data.ca\.crt}" > impersonation-proxy-self-signed-ca.pem.b64
+    
+# Tip: Put the contents of impersonation-proxy-self-signed-ca.pem.b64 into your copy buffer for a later step!
+```
+
+Install Pinniped’s Concierge:
+
+```shell
+$ kubectl apply \
+    --filename https://get.pinniped.dev/v0.25.0/install-pinniped-concierge-crds.yaml \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+
+$ kubectl apply \
+    --filename https://get.pinniped.dev/v0.25.0/install-pinniped-concierge-resources.yaml \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+```
+
+Now update the `CredentialIssuer` to use the Impersonation Proxy (which is disabled on kind by default):
+
+```shell
+$ kubectl edit credentialissuer pinniped-concierge-config \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+# Make sure that the spec has the following values:
+...
+  spec:
+    impersonationProxy:
+      externalEndpoint: impersonation-proxy-mtls.local
+      mode: enabled
+      service:
+        annotations:
+          service.beta.kubernetes.io/aws-load-balancer-connection-idle-timeout: "4000"
+        type: ClusterIP
+      tls:
+        certificateAuthorityData: # paste the contents of the file impersonation-proxy-self-signed-ca.pem.b64
+        secretName: impersonation-proxy-tls-serving-cert
+...
+# Now save and close the text editor
+
+# Confirm that the CredentialIssuer looks as expected
+$ kubectl get credentialissuers pinniped-concierge-config \
+    --output yaml \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+...
+  spec:
+    impersonationProxy:
+      externalEndpoint: impersonation-proxy-mtls.local
+      mode: enabled
+      service:
+        annotations:
+          service.beta.kubernetes.io/aws-load-balancer-connection-idle-timeout: "4000"
+        type: ClusterIP
+      tls:
+        certificateAuthorityData: foobar
+        secretName: impersonation-proxy-tls-serving-cert
+...
+
+# Confirm that the ClusterIP service was automatically created (may take a minute)
+$ kubectl get service pinniped-concierge-impersonation-proxy-cluster-ip \
+    --namespace pinniped-concierge \
+    --output yaml \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+
+# Configure a webhook authenticator to use the local-user-authenticator
+$ cat << EOF > concierge.webhookauthenticator.yaml
+apiVersion: authentication.concierge.pinniped.dev/v1alpha1
+kind: WebhookAuthenticator
+metadata:
+  name: local-user-authenticator
+spec:
+  endpoint: https://local-user-authenticator.local-user-authenticator.svc/authenticate
+  tls:
+    certificateAuthorityData: $(cat local-user-authenticator-ca.pem.b64)
+EOF
+
+# Create the webhook authenticator
+$ kubectl apply \
+    --filename concierge.webhookauthenticator.yaml \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+```
+
+Now deploy a Contour `HTTPProxy` ingress that fronts the `ClusterIP` service.
```

Comment:
Make it more obvious that we are trying to do TLS passthrough so that we can establish mTLS with the impersonation proxy in Pinniped's Concierge.

Reply:
done, see next push